A55-xds-stateful-session-affinity.md

```diff
@@ -0,0 +1,516 @@
+Once all assigned sessions have finished (or a timeout triggers), the backend
+application exits at which point the backend is now removed from the endpoints
+list (EDS). This design includes support for session draining.
+
+### Related Proposals: 
+
+* [A27: xDS-Based Global Load Balancing][A27]
+* [A42: xDS Ring Hash LB Policy][A42]
+
+## Proposal
+
+The goal of this design is to make it consistent with the Envoy implementation
+so it will interoperate with it. For example, one can create an xDS
+configuration for Envoy and the exact same configuration will work with
+proxyless gRPC. Similarly, an application using some cookie implementation
+that works with Envoy will continue to work with proxyless gRPC without any
+modifications.
+
+The design involves the following parts.
+
+### `XdsClient` to Include Endpoint Health Status
+
+The `EdsUpdate` struct from the `XdsClient` will be enhanced to include the
+[endpoint health status][eds-health-status] for each endpoint. `XdsClient`
+will also be modified to include endpoints in `DRAINING` state in the
+`EdsUpdate` reported to the watchers.
+
+### Listener and Route Configuration through xDS
+
+The feature is enabled and configured using the
+[cookie based stateful session extension][cookie-ext] which
+is an [http-filter][] in the `HttpConnectionManager` for the target service.
+A sample configuration is as follows:
+
+```textproto
+{ // HttpConnectionManager
+  http_filters: [
+    {
+      name: "envoy.filters.http.stateful_session"
+      typed_config: {
+        "@type": "type.googleapis.com/envoy.extensions.filters.http.stateful_session.v3.StatefulSession"
+        session_state: {
+          name: "envoy.http.stateful_session.cookie"
+          typed_config: {
+            "@type": "type.googleapis.com/envoy.extensions.http.stateful_session.cookie.v3.CookieBasedSessionState"
+            cookie: {
+              name: "global-session-cookie"
+              path: "/Package1.Service2/Method3"
+              ttl: "120s"
+            }
+          }
+        }
+      }
+    }
+  ]
+}
+```
+
+`name` is the name of the cookie used in this feature.
+
+`path` is the path for which this cookie is applied. **Note:** Stateful
+session affinity will not work if the cookie's path does not match the
+request's path. For example, if a given route matches requests with path
+`/service/foo` but the corresponding per-route filter config specifies a
+cookie path of `/service/bar`, then the cookie will never be set.
+
+`ttl` is the time-to-live for the cookie. It is set on the cookie but will
+not be enforced by gRPC.
+
+This new filter will be added to the HTTP Filter Registry and processed as
+described in [A39: xDS HTTP Filter Support][A39]. Specifically when this filter
+is present in the configuration, gRPC will create the appropriate client filter
+(aka interceptor in Java/Go) and install it in the channel to process data
+plane RPCs. We call this filter `CookieBasedStatefulSessionFilter`. It will
```

Comment:
We should just call it `StatefulSessionFilter`, without the `CookieBased` prefix.  In the future, we may add support for header-based state.

Reply:
"support for header-based state" will be a separate filter right? In that case does it make sense to keep the name as `CookieBasedStatefulSessionFilter` so we can have the other one called `HeaderBasedStatefulSessionFilter`

Reply:
No, I don't think we'd create a whole other filter to do this with headers.  We'd just add that as a new feature in the existing filter.

Please change this to refer to it as `StatefulSessionFilter`.  This is consistent with what Envoy calls it.

Reply:
done